File: Minimum Depth of Binary Tree.java
// Leetcode Problem 111: Minimum Depth of Binary Tree
// JAVA CODE
import java.util.*;

class Solution {

    public int minDepth(TreeNode root) {

        // Empty tree
        if (root == null) {
            return 0;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        int depth = 1;

        while (!queue.isEmpty()) {

            int size = queue.size();

            // Process one level
            for (int i = 0; i < size; i++) {

                TreeNode node = queue.poll();

                // First leaf found = minimum depth
                if (node.left == null && node.right == null) {
                    return depth;
                }

                // Add left child
                if (node.left != null) {
                    queue.offer(node.left);
                }

                // Add right child
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }

            depth++;
        }

        return 0;
    }
}

File: Minimum Depth of Binary Tree.cpp
// Leetcode Problem 111: Minimum Depth of Binary Tree
// C++ CODE
class Solution {
public:
    int minDepth(TreeNode* root) {

        // Empty tree
        if (root == nullptr) {
            return 0;
        }

        queue<TreeNode*> q;
        q.push(root);

        int depth = 1;

        while (!q.empty()) {

            int size = q.size();

            // Process current level
            for (int i = 0; i < size; i++) {

                TreeNode* node = q.front();
                q.pop();

                // First leaf found
                if (node->left == nullptr &&
                    node->right == nullptr) {
                    return depth;
                }

                // Add left child
                if (node->left != nullptr) {
                    q.push(node->left);
                }

                // Add right child
                if (node->right != nullptr) {
                    q.push(node->right);
                }
            }

            depth++;
        }

        return 0;
    }
};